### PR TITLE
Upgrade GraphQL.NET dependency to version 3.0.0

### PR DESCRIPTION
**Pull request by schabannes**
- Update GraphQL.Net to the latest ~preview~ version (3.0.0)
- Replace `UserContextWrapper` by new GraphQL.Net `UserContext` dictionary
- ~Rename IdGraphType name from 'ID' to 'Identity' to avoid clash with GraphQL.Net internal IdGraphType~
- Fix unit tests

### CODE REVIEW

**Comment by sungam3r** [comment]
3.0.0-preview-1271 is already not the latest version :)

File: src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs [hunk 1]
```diff
@@ -263,7 +263,7 @@ public GraphQLExecutor NewExecutor(IRequestDeserializer requestDeserializer = nu
 
         public string SerializeResult(ExecutionResult result)
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs]: Consider to make this method async as well.

File: src/GraphQL.Conventions/Adapters/Types/IdGraphType.cs [hunk 2]
```diff
@@ -16,7 +23,7 @@ public override object ParseValue(object value)
             var id = value?.ToString().StripQuotes();
             return NullableId(id);
         }
-
+
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Adapters/Types/IdGraphType.cs]: whitespaces

File: src/GraphQL.Conventions/GraphQL.Conventions.csproj [hunk 3]
```diff
@@ -2,9 +2,9 @@
 
   <PropertyGroup>
     <Description>GraphQL Conventions for .NET</Description>
-    <VersionPrefix>2.5.0</VersionPrefix>
+    <VersionPrefix>3.0.0</VersionPrefix>
     <Authors>Tommy Lillehagen</Authors>
-    <TargetFrameworks>netstandard1.6.1;net45</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0</TargetFrameworks>
     <DebugType>portable</DebugType>
```
Comment (sungam3r) [inline src/GraphQL.Conventions/GraphQL.Conventions.csproj:9]: I think `portable` is default value: https://github.com/dotnet/sdk/issues/1238

File: src/GraphQL.Conventions/GraphQL.Conventions.csproj [hunk 4]
```diff
@@ -29,21 +27,13 @@
     <LangVersion>7.3</LangVersion>
   </PropertyGroup>
 
-  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard1.6.1' ">
+  <ItemGroup>
+    <PackageReference Include="GraphQL" Version="3.0.0-preview-1271" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.2" />
```
Comment (sungam3r) [inline src/GraphQL.Conventions/GraphQL.Conventions.csproj]: No need for Newtonsoft.Json as GraphQL already includes it.

File: src/GraphQL.Conventions/Web/Response.cs [hunk 5]
```diff
@@ -40,7 +40,10 @@ public string Body
             {
                 if (string.IsNullOrWhiteSpace(_body) && ExecutionResult != null)
                 {
-                    _body = _writer.Write(ExecutionResult);
+                    _body = _writer.WriteToStringAsync(ExecutionResult)
+                        .ConfigureAwait(false)
+                        .GetAwaiter()
+                        .GetResult();
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Web/Response.cs]: Consider switch to async API.

File: test/Tests/Tests.csproj [hunk 6]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="15.0">
+﻿<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="15.0">
```
Comment (sungam3r) [inline test/Tests/Tests.csproj]: `ToolsVersion="15.0"` is not required

File: src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs (hunk 1 above)
Comment (schabannes) [reply src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs]: Added a SerializeResultAsync method. Kept the non-async one for backward compatibility.

File: Conventions.sln.DotSettings [hunk 7]
```diff
@@ -0,0 +1,2 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Comment (sungam3r) [inline Conventions.sln.DotSettings]: Is this file needed?

File: src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs [hunk 8]
```diff
@@ -261,10 +261,9 @@ public GraphQLExecutor NewExecutor(IRequestDeserializer requestDeserializer = nu
             return this;
         }
 
-        public string SerializeResult(ExecutionResult result)
-        {
-            return _documentWriter.WriteToStringAsync(result).GetAwaiter().GetResult();
-        }
+        public string SerializeResult(ExecutionResult result) => SerializeResultAsync(result).GetAwaiter().GetResult();
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs]: Consider to remove sync API or make it obsolete.

File: src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs [hunk 9]
```diff
@@ -306,22 +305,22 @@ private IGraphType ConstructOutputType(GraphTypeInfo typeInfo)
...
             };
             DeriveFields(typeInfo, graphType);
             return WrapNonNullableType(typeInfo, graphType);
         }
 
-        private object UnwrapObject(object obj)
+        private static object UnwrapObject(object obj)
         {
-            if (obj is INonNull)
+            if (obj is INonNull @null)
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs]:
```
if (obj is INonNull @null)
```
->
```
if (obj is INonNull nonNull)
```

otherwise it turns out weird

File: src/GraphQL.Conventions/Web/Response.cs [hunk 10]
```diff
@@ -26,33 +28,26 @@ public class Response
...
         public string QueryId => Request?.QueryId;
 
-        public ExecutionResult ExecutionResult { get; private set; }
+        public ExecutionResult ExecutionResult { get; }
 
-        public Validation.IValidationResult ValidationResult { get; private set; }
+        public Validation.IValidationResult ValidationResult { get; }
 
-        public string Body
+        public string Body => GetBodyAsync().GetAwaiter().GetResult();
```
Comment (sungam3r) [inline src/GraphQL.Conventions/Web/Response.cs]: Consider to remove sync API or mark it obsolete

Comment (schabannes) [reply src/GraphQL.Conventions/Web/Response.cs]: I wouldn't mind removing the sync ones. It would be a breaking change but as GraphQL.Net 3.0 already brings his lot of breaking changes it's not such a big deal and upgrading the code is quite straight forward. Do you have any version compatibility strategy ?

But if we do so it would be interesting to convert other method to async for consistency:
   - Response.Body: remove property (as per you other comment).
   - GraphQLEngine.SerializeResult: rename method (as per you other comment).
   - GraphQLEngine.Execute: rename to ExecuteAsync
   - IGraphQLExecutor.Execute: rename to ExecuteAsync
   - IGraphQLExecutor.Validate: convert to Async
   - IRequestHandler.ProcessRequest: rename to ProcessRequestAsync
   - IRequestHandler.Validate: convert to Async
   - IRequestHandler.DescribeSchema: convert to Async

What do you think?

Comment (sungam3r) [reply src/GraphQL.Conventions/Web/Response.cs]: > Do you have any version compatibility strategy ?

Well, graphql-dotnet version 3 (next major) is already backward incompatible in many places, but I don’t see a problem in that. After the release, all dependent packages will need to release their major versions as well, that's all.

> What do you think?

I think that changing the signature of some methods to the asynchronous version is perfectly normal. See graphql-dotnet/graphql-dotnet#1371

Comment (sungam3r) [reply src/GraphQL.Conventions/Web/Response.cs]: I note that the server project is also in the preview version using the preview of the base library. I can’t say that the preview version of graphql-dotnet is less stable than 2.4, rather the opposite.

Comment (schabannes) [reply src/GraphQL.Conventions/Web/Response.cs]: I created a separate PR #195

File: src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs (hunk 1 above)
Comment (schabannes) [reply src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs]: see #195

File: src/GraphQL.Conventions/Web/Response.cs (hunk 5 above)
Comment (schabannes) [reply src/GraphQL.Conventions/Web/Response.cs]: see #195

File: Conventions.sln.DotSettings (hunk 7 above)
Comment (schabannes) [reply Conventions.sln.DotSettings]: No, I removed it

File: src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs (hunk 8 above)
Comment (schabannes) [reply src/GraphQL.Conventions/Adapters/Engine/GraphQLEngine.cs]: see #195

**Comment by yokan87** [comment]
@tlil @sungam3r @schabannes 
Hi,
is it planned to continue this pull request, or are there any other ways to use this library together with the current GraphQL-Dotnet 3.0.0 preview?
In the Readme of the graphql-dotnet library it is mentioned that the 2.4.0 has many known issues and as I have some problems with the reliability of subscriptions, I wanted to proof if the latest preview of the libraries would fix it.

Kind regards

**Comment by tlil** [comment]
Hi @jdomnick - yes, I'm planning to get back to this soon. Feel free to have a stab at it in the meantime ;-)

**Comment by schabannes** [comment]
Includes #206

**Comment by tlil** [comment]
Thanks, @schabannes - I'm getting 3 unit test failures when running locally. Have you looked at those yet? Will review more closely over the weekend.

**Comment by schabannes** [comment]
@till, I fixed the test. Something was wrong in the last merge from master. I did review manually and it looks alright now, but an extra pair of eyes is always better.

File: test/GraphQL.Conventions.Tests/Adapters/Engine/ExposingExceptionsTests.cs [hunk 11]
```diff
@@ -12,36 +12,33 @@ public class ExposingExceptionsTests : TestBase
...
             GraphQL.ExecutionResult result = await engine
                 .NewExecutor()
                 .WithQueryString("query { queryData }")
-                .Execute();
-
-            result.ExposeExceptions.ShouldBeFalse("By default exposing exceptions should be disabled.");
+                .ExecuteAsync();
 
-            var message = JObject.FromObject(result)["errors"].First.Value<string>("message");
+            var message = JObject.FromObject(result)["Errors"].First.Value<string>("Message");
```
Comment (tlil) [inline test/GraphQL.Conventions.Tests/Adapters/Engine/ExposingExceptionsTests.cs]: I would expect these to be `"errors"` and `"message"` still.

File: .github/workflows/dotnetcore.yml [hunk 12]
```diff
@@ -26,14 +26,13 @@ jobs:
       - name: Generate NuGet package
         run: |
           dotnet pack -c Release
-
+
```
Comment (tlil) [inline .github/workflows/dotnetcore.yml]:
```suggestion

```

File: .github/workflows/dotnetcore.yml [hunk 13]
```diff
@@ -26,14 +26,13 @@ jobs:
...
           dotnet pack -c Release
-
+          
       - name: Install NuGet client
         uses: warrenbuckley/Setup-Nuget@v1
 
       - name: Add private GitHub registry to NuGet
         run: |
           nuget sources add -name "GPR" -Source https://nuget.pkg.github.com/graphql-dotnet/index.json -Username tlil -Password ${{ secrets.GITHUB_TOKEN }}
-
```
Comment (tlil) [inline .github/workflows/dotnetcore.yml]: *nit*: Leave empty line

File: src/GraphQL.Conventions/Adapters/Engine/GraphQLExecutor.cs [hunk 14]
```diff
@@ -74,7 +78,6 @@ public IGraphQLExecutor<ExecutionResult> WithRootObject(object rootValue)
             _rootObject = rootValue;
             return this;
         }
-
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/Engine/GraphQLExecutor.cs]: *nit*: Leave blank line, please.

File: src/GraphQL.Conventions/Adapters/Engine/GraphQLExecutor.cs [hunk 15]
```diff
@@ -65,7 +65,11 @@ public IGraphQLExecutor<ExecutionResult> WithOperationName(string operationName)
 
         public IGraphQLExecutor<ExecutionResult> WithInputs(Dictionary<string, object> inputs)
         {
-            _inputs = inputs != null ? new Inputs(inputs) : new Inputs();
+            return WithInputs(new Inputs(inputs ?? new Dictionary<string, object>()));
+        }
+        public IGraphQLExecutor<ExecutionResult> WithInputs(Inputs inputs)
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/Engine/GraphQLExecutor.cs:71]:
```suggestion

        public IGraphQLExecutor<ExecutionResult> WithInputs(Inputs inputs)
```

File: src/GraphQL.Conventions/Adapters/Engine/GraphQLExecutor.cs [hunk 16]
```diff
@@ -133,9 +136,9 @@ public IGraphQLExecutor<ExecutionResult> DisableProfiling()
             return this.EnableProfiling(false);
         }
 
-        public async Task<ExecutionResult> Execute() =>
+        public async Task<ExecutionResult> ExecuteAsync() =>
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/Engine/GraphQLExecutor.cs]: Seeing this now, I realise we can skip the `async`, `await` and `ConfigureAwait()`. :-) Simply:
```suggestion
        public Task<ExecutionResult> ExecuteAsync() => _engine.ExecuteAsync(...);
```

File: src/GraphQL.Conventions/Adapters/Engine/IGraphQLExecutor.cs [hunk 17]
```diff
@@ -15,7 +15,9 @@ public interface IGraphQLExecutor<TResult>
 
         IGraphQLExecutor<TResult> WithOperationName(string operationName);
 
-        IGraphQLExecutor<TResult> WithInputs(Dictionary<string, object> inputs);
+        IGraphQLExecutor<TResult> WithInputs(Inputs inputs);
+
+        IGraphQLExecutor<ExecutionResult> WithInputs(Dictionary<string, object> inputs);
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/Engine/IGraphQLExecutor.cs]:
```suggestion
        IGraphQLExecutor<TResult> WithInputs(Dictionary<string, object> inputs);
```

File: src/GraphQL.Conventions/Adapters/Resolvers/FieldResolver.cs [hunk 18]
```diff
@@ -85,5 +84,7 @@ private object GetSource(GraphFieldInfo fieldInfo, IResolutionContext context)
             source = Unwrapper.Unwrap(source);
             return source;
         }
+
+
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/Resolvers/FieldResolver.cs]:
```suggestion
```

File: src/GraphQL.Conventions/Execution/UserContextExtensions.cs [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+        public static IDependencyInjector GetDependencyInjector(this IProvideUserContext context) => GetValue<IDependencyInjector>(context.UserContext);
+        public static void SetDependencyInjector(this IProvideUserContext context, IDependencyInjector value) => SetValue(context.UserContext, value);
+
+        private static T GetValue<T>(IDictionary<string, object> dictionary)
+        {
+            var key = typeof(T).FullName;
+            return dictionary != null && dictionary.ContainsKey(key) 
+                ? (T) dictionary[key] 
+                : default;
```
Comment (tlil) [inline src/GraphQL.Conventions/Execution/UserContextExtensions.cs]: `TryGetValue()` ?

File: src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs [hunk 20]
```diff
@@ -12,7 +12,7 @@ public static class TypeNames
 
         public const string Boolean = "Boolean";
 
-        public const string Id = "ID";
+        public const string Identity = "Identity";
```
Comment (tlil) [inline src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs]: Why this change?

https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Types/Scalars/IdGraphType.cs

File: src/GraphQL.Conventions/Web/Response.cs [hunk 21]
```diff
@@ -26,30 +28,24 @@ public class Response
...
-                    _body = _writer.Write(ExecutionResult);
-                }
-                return _body;
-            }
-            internal set
-            {
-                _body = value;
-            }
+            if (string.IsNullOrWhiteSpace(_body) && ExecutionResult != null)
+                _body = await _writer.WriteToStringAsync(ExecutionResult);
```
Comment (tlil) [inline src/GraphQL.Conventions/Web/Response.cs]: `ConfigureAwait(false)`?

File: test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs [hunk 22]
```diff
@@ -76,51 +77,56 @@ public async Task Will_Provide_Path_And_Code_For_Errors_In_Array_Fields()
...
             result.Data.ShouldHaveFieldWithValue("errorWithData", null);
-
             result.Errors.ShouldNotBeNull();
             result.Errors.Count.ShouldEqual(1);
 
-            result.Errors.First().Message.ShouldEqual("Test error.");
-            result.Errors.First().Data["someKey"].ShouldEqual("someValue");
+            var error = result.Errors.First();
+            error.InnerException.ShouldNotBeNull();
+            error.InnerException.InnerException.ShouldNotBeNull();
```
Comment (tlil) [inline test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: What's the cause for the nesting here...? What's changed, and how does this error actually manifest itself in the response payload now?

File: test/GraphQL.Conventions.Tests/Attributes/MetaData/FieldTypeMetaDataAttributeTests.cs [hunk 23]
```diff
@@ -137,12 +137,11 @@ public TestCustomAttribute(string permission)
 
     public class TestValidation : IValidationRule
     {
-        public INodeVisitor Validate(ValidationContext context)
+        public async Task<INodeVisitor> ValidateAsync(ValidationContext context)
```
Comment (tlil) [inline test/GraphQL.Conventions.Tests/Attributes/MetaData/FieldTypeMetaDataAttributeTests.cs]: No need to add `async` and `await` here? Just `return Task.FromResult(...);` ?

**Review by tlil — CHANGES_REQUESTED** [review]
Thanks for this. A few more comments and questions, sorry :-)

File: test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs (hunk 22 above)
Comment (schabannes) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: The source exception is wrapped twice in graphql-dotnet. Once by the .net framework (methodInfo.Invoke) and once by the ExecutionStrategy (ProcessNodeUnhandledException).

Comment (tlil) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: Gotcha. Sounds like we can remove at least one level of indirection there, no?

Comment (schabannes) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: Not to access the data attached to the exception. We could unwrap the exception in GraphQLEngine (via an ErrorTransformer maybe) but it would probably be best to handle that in Graphql.net directly as it is the one who wraps it.

Comment (tlil) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: Ah, I see - let me take it for a spin later and see for myself what the current behaviour is. Will get back to you.

Comment (Shane32) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: I was writing code earlier today to prevent `MethodInfo.Invoke` from wrapping an exception.  I explored and tested `MethodInfo.Invoke` and `Delegate.DynamicInvoke`, and explored using `ExceptionDispatchInfo.Capture`.  I settled on compiling an expression so there is no need to rethrow the exception, using a cache to reduce the impact of dynamically compiling the expression.  I can paste the code in here for you to use @schabannes @tlil .  Using `ExceptionDispatchInfo.Capture` is probably the best alternative choice if this seems too complex.

Comment (tlil) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: Yes, please @Shane32 - if you could share, I'd be happy to take a look.

Comment (Shane32) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]:
```csharp
    class MethodInvokeTests
    {
        public async Task RunAsync()
        {
            Console.WriteLine($"StaticTest1: {Invoke(GetMethodInfo("StaticTest1"), null, null)}");
            Console.WriteLine($"StaticTest2: {Invoke(GetMethodInfo("StaticTest2"), null, null)}");
            Console.WriteLine($"StaticTest3: {Invoke(GetMethodInfo("StaticTest3"), null, null)}");
            Console.WriteLine($"InstanceTest1: {Invoke(GetMethodInfo("InstanceTest1"), this, null)}");
            Console.WriteLine($"InstanceTest2: {Invoke(GetMethodInfo("InstanceTest2"), this, null)}");
            Console.WriteLine($"InstanceTest3: {Invoke(GetMethodInfo("InstanceTest3"), this, null)}");
            var guid = Guid.NewGuid();
            var stream = new System.IO.MemoryStream();
            var value = (int?)35;
            Console.WriteLine($"StaticWithParams: {Invoke(GetMethodInfo("StaticWithParams"), null, new object[] { guid, stream, value })}");
            Console.WriteLine($"InstanceWithParams: {Invoke(GetMethodInfo("InstanceWithParams"), this, new object[] { guid, stream, value })}");
        }

        private MethodInfo GetMethodInfo(string name) => typeof(MethodInvokeTests).GetMethod(name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
        private static int StaticTest1() => 2;
        private static string StaticTest2() => "hello";
        private static void StaticTest3() => Console.WriteLine("ran StaticTest3");
        private int InstanceTest1() => 3;
        private string InstanceTest2() => "hello2";
        private void InstanceTest3() => Console.WriteLine("ran InstanceTest3");
        private static string StaticWithParams(Guid guid, Stream stream, int? value) => $"Static Received {guid} {stream} {value}";
        private string InstanceWithParams(Guid guid, Stream stream, int? value) => $"Instance Received {guid} {stream} {value}";

        private static ConcurrentDictionary<MethodInfo, Func<object, object[], object>> _methodDictionary = new ConcurrentDictionary<MethodInfo, Func<object, object[], object>>();
        public static object Invoke(MethodInfo methodInfo, object instance, params object[] arguments)
        {
            //just good practice
            if (methodInfo == null)
                throw new ArgumentNullException(nameof(methodInfo));
            //the following check is ignored by default
            if (methodInfo.IsStatic && instance != null)
                throw new ArgumentException("Instance must be null for static methods", nameof(instance));
            //the lambda ignores extra arguments so the following check is necessary
            if (methodInfo.GetParameters().Length != (arguments?.Length ?? 0))
                throw new ArgumentException("Invalid number of arguments for this method", "arguments");
            //type errors will be thrown as InvalidCastException inside the lambda
            //null instance (for non-static methods) will be thrown as NullReferenceException inside the lambda

            var lambda = _methodDictionary.GetOrAdd(methodInfo, (methodInfo) => {
                var instanceParameter = Expression.Parameter(typeof(object));
                var argumentsParameter = Expression.Parameter(typeof(object[]));
                var instanceExpression = methodInfo.IsStatic ? null : Expression.Convert(instanceParameter, methodInfo.DeclaringType);
                var parameters = methodInfo.GetParameters().Select((param, index) => {
                    return Expression.Convert(Expression.ArrayAccess(argumentsParameter, Expression.Constant(index)), param.ParameterType);
                });
                var call = Expression.Call(instanceExpression, methodInfo, parameters);
                Expression body;
                if (methodInfo.ReturnType == typeof(void)) {
                    body = Expression.Block(new Expression[] {
                        call,
                        Expression.Constant(null, typeof(object))
                    });
                }
                else {
                    body = Expression.Convert(call, typeof(object));
                }
                var lambda = Expression.Lambda<Func<object, object[], object>>(body, instanceParameter, argumentsParameter);
                return lambda.Compile();
            });

            return lambda(instance, arguments);
        }
    }
```

Comment (Shane32) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: And here's the simple version, which does almost the same thing, catching and rethrowing the inner exception while maintaining the stack trace:
```csharp
        public static object Invoke2(MethodInfo methodInfo, object instance, params object[] arguments)
        {
            try {
                return methodInfo.Invoke(instance, arguments);
            }
            catch (TargetInvocationException ex) {
                ExceptionDispatchInfo.Capture(ex.InnerException).Throw();
                return null; //does not run, but intellisense doesn't know that
            }
        }
```
Note that in theory, this could catch a `TargetInvocationException` that occurred within the method.

In terms of performance, the former option is probably slower for the first-time execution, as it has to compile an expression into a dynamically generated assembly at runtime.  But it should be just as fast after the first time; perhaps even faster.  It would be interesting to see performance benchmarks and see which is faster in the long run.

Personally, I don't like catching and rethrowing exceptions, as when debugging an unhandled exception, Visual Studio invariably seems to highlight the line of code that's rethrowing the exception, not where it originally occurred.  Also, the stack traces don't contain a "--- End of stack trace from previous location where exception was thrown ---" message.  Besides those reasons, the simple alternative is probably perfectly adequate.

Comment (Shane32) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: I've not been following the discussion so I'm not sure where this is used, and so perhaps my code could be tweaked for better performance given the usage.  Also I could have commented it better and added xml comments...  (Actually, xml comments could be copied and pasted from MethodInfo.Invoke, I bet.)

File: test/GraphQL.Conventions.Tests/Adapters/Engine/ExposingExceptionsTests.cs (hunk 11 above)
Comment (schabannes) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ExposingExceptionsTests.cs]: It is, to some extend, another breaking change from graphQL.Net. As they extract Newtonsoft to a separate library they removed the JsonConverterAttribute from the ExecutionResults object. Here the result is serialized using the default value from Newtonsoft which is PascalCase.
But now I'm looking at those tests they are not relevant anymore as they were testing the 'ExposingExceptions' option which has been removed in graphQL.Net. It is replaced by an IErrorProvider, this isn't supported by Conventions right now.
For now I would simply remove those tests.

File: src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs (hunk 20 above)
Comment (schabannes) [reply src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs]: There was a clash between both IdGraphType (the conventions one and the graphQL.Net one). However I did that more than a year ago and I can't remember why I did that. Maybe the Conventions.IdGraphType is redundant now?
I will revert the changes around this so we can move forward with this PR and we can do more test around the IdGraphType.

File: src/GraphQL.Conventions/Execution/UserContextExtensions.cs (hunk 19 above)
Comment (schabannes) [reply src/GraphQL.Conventions/Execution/UserContextExtensions.cs]: Because TryGetValue use an out param, it doesn't really make the code shorter or more readable.

File: test/GraphQL.Conventions.Tests/Adapters/Engine/ExposingExceptionsTests.cs (hunk 11 above)
Comment (sungam3r) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ExposingExceptionsTests.cs]: > they removed the JsonConverterAttribute from the `ExecutionResult`

Yes. All serialization settings were moved to appropriate external packages. `ExecutionResult` itself is serialization agnostic.

File: src/GraphQL.Conventions/Execution/UserContextExtensions.cs (hunk 19 above)
Comment (sungam3r) [reply src/GraphQL.Conventions/Execution/UserContextExtensions.cs]: I makes code to perform only one dictionary lookup instead of two.

File: src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs (hunk 20 above)
Comment (schabannes) [reply src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs]: The error is:
Unable to register GraphType 'GraphQL.Conventions.Adapters.Types.IdGraphType' with the name 'ID' the name 'ID' is already registered to 'GraphQL.Types.IdGraphType'
Any idea what would be the best way to handle this ?

Comment (sungam3r) [reply src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs]: https://github.com/graphql-dotnet/graphql-dotnet/issues/1952

File: src/GraphQL.Conventions/Execution/UserContextExtensions.cs (hunk 19 above)
Comment (Shane32) [reply src/GraphQL.Conventions/Execution/UserContextExtensions.cs]: Agree, one lookup is better than two. You can still do it in one line of code, nearly the same as it’s written now

File: src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs (hunk 20 above)
Comment (schabannes) [reply src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs]: I looked at this issue differently, Instead of having an overload of IdGraphType, I added a TypeConverter on the Id class (and support in the primitive wrapper).

File: test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs (hunk 22 above)
Comment (schabannes) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: Nice. I'll try to wrap that in a PR on the graphql-dotnet main repo.

Comment (Shane32) [reply test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs:125]: @schabannes Actually, the main repo already unwraps exceptions triggered by MethodInfo.Invoke.  It does not, however, retain the proper stack trace or unwrap exceptions triggered by ConstructorInfo.Invoke.  I've proposed slight changes in https://github.com/graphql-dotnet/graphql-dotnet/pull/1988 .  I've also put together PR #210 which contains the necessary changes to this project.  As this project relies heavily on `MethodInfo.Invoke`, I've included the more exotic code so that exceptions are not rethrown at all, along with extensive tests.

File: src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs (hunk 20 above)
Comment (tlil) [reply src/GraphQL.Conventions/Types/Descriptors/TypeNames.cs]: Thanks :+1:

**Review by tlil — APPROVED** [review]

**Comment by tlil** [comment]
Thank you, all! This is now published to NuGet: https://www.nuget.org/packages/GraphQL.Conventions/3.0.0

https://github.com/graphql-dotnet/conventions/releases/tag/v3.0.0
